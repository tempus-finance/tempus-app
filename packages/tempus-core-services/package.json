{
  "name": "tempus-core-services",
  "version": "0.0.1",
  "private": true,
  "description": "Tempus core Services",
  "keywords": [
    "typescript",
    "boilerplate",
    "esbuild"
  ],
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "scripts": {
    "lint": "eslint src/ --ext .js,.jsx,.ts,.tsx",
    "test": "jest",
    "clean": "rm -rf dist build package",
    "ts-node": "ts-node",
    "docs": "typedoc --entryPoints src/index.ts",
    "build": "npm run clean && tsc -p tsconfig.json",
    "build-all": "npm run clean && npm run build && npm run esbuild-node && npm run esbuild-node",
    "esbuild-node": "esbuild src/index.ts --bundle --platform=node --minify --sourcemap=external --outfile=dist/esbuild/index.js",
    "esbuild-node:dev": "esbuild src/index.ts --bundle --platform=node --sourcemap=external --outfile=dist/esbuild/index.js",
    "esbuild-node:watch": "esbuild src/index.ts --bundle --platform=node --watch --sourcemap=external --outfile=dist/esbuild/index.js"
  },
  "dependencies": {
    "@ethersproject/providers": "^5.5.2",
    "@types/jest": "^24.9.1",
    "@types/node": "^12.20.15",
    "axios": "^0.22.0",
    "date-fns": "^2.28.0",
    "ethers": "^5.4.6"
  },
  "devDependencies": {
    "esbuild": "^0.11.11",
    "eslint": "^7.22.0",
    "jest": "^26.6.3",
    "ts-jest": "^26.5.4",
    "ts-node": "^9.1.1",
    "typedoc": "^0.20.35"
  },
  "peerDependencies": {
    "@typescript-eslint/eslint-plugin": "^4.28.1",
    "@typescript-eslint/parser": "^4.28.1",
    "eslint-config-airbnb-typescript": "^12.3.1"
  }
}
