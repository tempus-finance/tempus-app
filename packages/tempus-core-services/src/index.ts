import ERC20ABI from './abi/ERC20.json';
import VaultABI from './abi/Vault.json';

export { ERC20ABI, VaultABI };

export type { ERC20 } from './abi';

export type {
  AvailableToDeposit,
  Chain,
  ChainConfig,
  Config,
  DynamicTempusPool,
  PoolShares,
  ProtocolDisplayName,
  ProtocolName,
  Ticker,
  NativeTokenTicker,
  TempusPool,
  TokenPrecision,
  TokenTypePrecision,
  YearnData,
  YearnDataApy,
  YearnDataToken,
} from './interfaces';

export type {
  DepositedEvent,
  ERC20TokenService,
  PoolBalanceChangedEvent,
  PoolBalanceChangedEventListener,
  RedeemedEvent,
  TempusAMMJoinKind,
  TempusAMMExitKind,
  TempusAMMService,
  TempusControllerService,
  TempusPoolService,
  TransferEventListener,
  VaultService,
  StatisticsService,
  StorageService,
  SwapEvent,
} from './services';

export type { Numberish } from './datastructures';

export { Decimal, ZERO, ONE, DEFAULT_DECIMAL_PRECISION } from './datastructures';

export { SwapKind } from './services';

export {
  getChainlinkFeed,
  getCoingeckoRate,
  getDefaultProvider,
  getERC20TokenService,
  initServices,
  getServices,
  getDefinedServices,
} from './services';

export {
  NumberUtils,
  DecimalUtils,
  capitalize,
  chainIdToChainName,
  chainIdHexToChainName,
  chainNameToHexChainId,
  chainToTicker,
  decreasePrecision,
  div18f,
  getCookie,
  getLibrary,
  getPastDaysNumber,
  getProvider,
  getProviderFromSignerOrProvider,
  getRangeFrom,
  getTokenPrecision,
  increasePrecision,
  isZeroString,
  mul18f,
  prettifyChainName,
  prettifyChainNameLong,
  rayToDai,
  reverseString,
  shortenAccount,
  wadToDai,
  wait,
  prettifyProtocolName,
} from './utils';

export * from './constants';
