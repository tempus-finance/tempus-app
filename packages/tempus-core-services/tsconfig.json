{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "declaration": true,

    "importHelpers": true,
    "experimentalDecorators": true,
    "sourceMap": true,
    "outDir": "./dist/",
    "types": ["node", "jest"]
  },
  "include": ["src/**/*.ts"],
  "exclude": ["node_modules"]
}
