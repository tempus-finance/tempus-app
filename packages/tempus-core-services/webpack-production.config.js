const path = require('path');
const webpack = require('webpack');

require('dotenv').config({ path: './.env.production' });

module.exports = {
  entry: './src/index.ts',
  module: {
    rules: [
      {
        test: /\.tsx?$/,
        use: 'ts-loader',
        exclude: [/node_modules/],
      },
    ],
  },
  resolve: {
    extensions: ['.ts', '.js'],
  },
  output: {
    path: path.resolve(__dirname, 'dist'),
    publicPath: '',
    filename: 'index.js',
    library: 'tempus-core-services',
    libraryTarget: 'umd',
  },
  mode: 'production',
  devtool: 'eval-source-map',
  externals: ['ethers', 'rxjs', 'date-fns', '@ethersproject/providers'],
  plugins: [
    new webpack.DefinePlugin({
      'process.env': JSON.stringify(process.env),
    }),
  ],
};
