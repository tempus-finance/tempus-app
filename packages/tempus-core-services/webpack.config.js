// Webpack uses this to work with directories
const path = require('path');
const webpack = require('webpack');

require('dotenv').config({ path: '../tempus-client_v3/.env.development' });

// This is the main configuration object.
// Here, you write different options and tell Webpack what to do
module.exports = {
  // Path to your entry point. From this file Webpack will begin its work
  entry: './src/index.ts',

  module: {
    rules: [
      {
        test: /\.tsx?$/,
        use: 'ts-loader',
        exclude: [/node_modules/],
        include: [path.resolve(__dirname, 'src')],
      },
    ],
  },
  resolve: {
    extensions: ['.ts', '.js'],
  },

  // Path and filename of your result bundle.
  // Webpack will bundle all JavaScript into this file
  output: {
    path: path.resolve(__dirname, 'dist'),
    publicPath: '',
    filename: 'index.js',
    library: 'tempus-core-services',
    libraryTarget: 'umd',
  },

  // Default mode for Webpack is production.
  // Depending on mode Webpack will apply different things
  // on the final bundle. For now, we don't need production's JavaScript
  // minifying and other things, so let's set mode to development
  mode: 'development',

  devtool: 'eval-source-map',

  watch: true,

  plugins: [
    new webpack.DefinePlugin({
      'process.env': JSON.stringify(process.env),
    }),
  ],
};
