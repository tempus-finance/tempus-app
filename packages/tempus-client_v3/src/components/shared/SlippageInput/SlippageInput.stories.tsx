import { ComponentStory } from '@storybook/react';
import { FC, useCallback, useState } from 'react';
import { Decimal } from 'tempus-core-services';
import SlippageInput, { SlippageInputProps } from './SlippageInput';
import Typography from '../Typography';

export default {
  title: 'SlippageInput',
  component: SlippageInput,
  argTypes: {},
};

const style = {
  background: 'rgba(0, 0, 0, 0.1)',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  padding: '10px',
};

const Wrapper: FC<SlippageInputProps> = props => {
  const [percentage, setPercentage] = useState<Decimal>(props.percentage ?? new Decimal(0));
  const [isAuto, setIsAuto] = useState<boolean>(props.isAuto ?? false);
  const onPercentageUpdate = useCallback((val: Decimal) => setPercentage(val), []);
  const onAutoUpdate = useCallback((val: boolean) => setIsAuto(val), []);
  return (
    <div>
      <SlippageInput
        {...props}
        percentage={percentage}
        isAuto={isAuto}
        onPercentageUpdate={onPercentageUpdate}
        onAutoUpdate={onAutoUpdate}
      />
      <hr />
      <Typography variant="body-secondary">Percentage: {percentage?.toString()}</Typography>
      <Typography variant="body-secondary">set Auto: {isAuto.toString()}</Typography>
    </div>
  );
};

const Template: ComponentStory<typeof SlippageInput> = args => (
  <div style={style}>
    <Wrapper {...args} />
  </div>
);

export const SlippageInputNormal = Template.bind({});
SlippageInputNormal.args = {
  percentage: new Decimal(0.02),
  isAuto: false,
  error: undefined,
  disabled: false,
};

export const SlippageInputAuto = Template.bind({});
SlippageInputAuto.args = {
  percentage: new Decimal(0.02),
  isAuto: true,
  error: undefined,
  disabled: false,
};

export const SlippageInputError = Template.bind({});
SlippageInputError.args = {
  percentage: new Decimal(0.02),
  isAuto: false,
  error: 'cannot exceed 100%',
  disabled: false,
};

export const SlippageInputDisabled = Template.bind({});
SlippageInputDisabled.args = {
  percentage: new Decimal(0.02),
  isAuto: false,
  error: undefined,
  disabled: true,
};
