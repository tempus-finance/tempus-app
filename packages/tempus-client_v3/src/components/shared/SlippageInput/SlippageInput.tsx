import { FC, useCallback, useEffect, useMemo, useState } from 'react';
import { Decimal, DecimalUtils } from 'tempus-core-services';
import ButtonWrapper from '../ButtonWrapper';
import TextInput from '../TextInput';
import Typography, { TypographyColor } from '../Typography';

import './SlippageInput.scss';

export interface SlippageInputProps {
  percentage: Decimal;
  isAuto: boolean;
  error?: string;
  disabled?: boolean;
  onPercentageUpdate?: (value: Decimal) => void;
  onAutoUpdate?: (value: boolean) => void;
}

const SlippageInput: FC<SlippageInputProps> = props => {
  const { percentage, isAuto, error, disabled, onPercentageUpdate, onAutoUpdate } = props;
  const [placeholder, setPlaceholder] = useState<string>(percentage.mul(100).toRounded(2));
  const [value, setValue] = useState<string>('');
  const [focused, setFocused] = useState<boolean>(false);

  useEffect(() => setPlaceholder(DecimalUtils.formatPercentage(percentage, 2)), [percentage]);

  const handleTextboxFocus = useCallback(() => setFocused(true), []);
  const handleTextboxBlur = useCallback(() => setFocused(false), []);
  const handleAutoClick = useCallback(() => {
    onAutoUpdate?.(true);
  }, [onAutoUpdate]);
  const handleDebounceChange = useCallback(
    (inputValue: string) => {
      onPercentageUpdate?.(Decimal.parse(inputValue, 0).div(100));
      onAutoUpdate?.(false);
    },
    [onPercentageUpdate, onAutoUpdate],
  );

  const autoButton = useMemo(() => {
    let color: TypographyColor | undefined;
    if (disabled) {
      color = 'text-disabled';
    } else if (isAuto) {
      color = 'text-primary-inverted';
    }

    return (
      <ButtonWrapper selected={isAuto} disabled={disabled} onClick={handleAutoClick}>
        <Typography variant="body-primary" weight="regular" color={color}>
          Auto
        </Typography>
      </ButtonWrapper>
    );
  }, [isAuto, disabled, handleAutoClick]);

  return (
    <div className="tc__slippage-input">
      <TextInput
        placeholder={placeholder}
        value={focused ? value : ''}
        pattern="[0-9,]*[.]?[0-9]{0,2}"
        debounce
        error={error}
        disabled={disabled}
        endAdornment={autoButton}
        onChange={setValue}
        onFocus={handleTextboxFocus}
        onBlur={handleTextboxBlur}
        onDebounceChange={handleDebounceChange}
      />
    </div>
  );
};

export default SlippageInput;
