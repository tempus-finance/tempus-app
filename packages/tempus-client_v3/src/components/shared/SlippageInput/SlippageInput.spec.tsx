import { act, fireEvent, render } from '@testing-library/react';
import { Decimal, DecimalUtils } from 'tempus-core-services';
import { FC, useState } from 'react';
import SlippageInput, { SlippageInputProps } from './SlippageInput';

const mockOnPercentageUpdate = jest.fn();
const mockOnAutoUpdate = jest.fn();

const defaultProps: SlippageInputProps = {
  percentage: new Decimal(0.02),
  isAuto: false,
  error: undefined,
  disabled: false,
  onPercentageUpdate: mockOnPercentageUpdate,
  onAutoUpdate: mockOnAutoUpdate,
};

const Wrapper: FC<SlippageInputProps> = props => {
  const [percentage, setPercentage] = useState<Decimal>(props.percentage ?? new Decimal(0));
  const [isAuto, setIsAuto] = useState<boolean>(props.isAuto ?? false);
  mockOnPercentageUpdate.mockImplementation((val: Decimal) => setPercentage(val));
  mockOnAutoUpdate.mockImplementation((val: boolean) => setIsAuto(val));
  return <SlippageInput {...props} percentage={percentage} isAuto={isAuto} />;
};

const subject = (props: SlippageInputProps) => render(<Wrapper {...props} />);

describe('SlippageInput', () => {
  it('renders a slippage input', () => {
    const { container, getByRole } = subject(defaultProps);

    const slippageInput = container.querySelector('.tc__slippage-input');
    const input = getByRole('textbox');
    const button = getByRole('button');

    expect(slippageInput).not.toBeNull();
    expect(input).not.toBeNull();
    expect(button).not.toBeNull();
    expect(input).toHaveAttribute('placeholder', DecimalUtils.formatPercentage(defaultProps.percentage, 2));

    expect(slippageInput).toMatchSnapshot();
  });

  it('renders a slippage input set to auto', () => {
    const props = { ...defaultProps, isAuto: true };
    const { container, getByRole } = subject(props);

    const slippageInput = container.querySelector('.tc__slippage-input');
    const input = getByRole('textbox');
    const button = getByRole('button');

    expect(slippageInput).not.toBeNull();
    expect(input).not.toBeNull();
    expect(button).not.toBeNull();
    expect(button.getAttribute('data-selected')).toEqual('true');
    expect(input).toHaveAttribute('placeholder', DecimalUtils.formatPercentage(props.percentage, 2));

    expect(slippageInput).toMatchSnapshot();
  });

  it('renders a disabled slippage input', () => {
    const props = { ...defaultProps, disabled: true };
    const { container, getByRole } = subject(props);

    const slippageInput = container.querySelector('.tc__slippage-input');
    const input = getByRole('textbox');
    const button = getByRole('button');

    expect(slippageInput).not.toBeNull();
    expect(input).not.toBeNull();
    expect(button).not.toBeNull();
    expect(input).toHaveAttribute('placeholder', DecimalUtils.formatPercentage(props.percentage, 2));
    expect(input.getAttribute('disabled')).not.toBeNull();

    expect(slippageInput).toMatchSnapshot();
  });

  it('renders a slippage input with error', () => {
    const props = { ...defaultProps, error: 'cannot exceed 100%' };
    const { container, getByRole, queryByText } = subject(props);

    const slippageInput = container.querySelector('.tc__slippage-input');
    const input = getByRole('textbox');
    const button = getByRole('button');
    const error = queryByText(props.error);

    expect(slippageInput).not.toBeNull();
    expect(input).not.toBeNull();
    expect(button).not.toBeNull();
    expect(error).not.toBeNull();
    expect(input).toHaveAttribute('placeholder', DecimalUtils.formatPercentage(props.percentage, 2));

    expect(slippageInput).toMatchSnapshot();
  });

  it('typing in slippage input will trigger onPercentageUpdate() and call onAutoUpdate(false) with debounce', () => {
    jest.useFakeTimers();

    const { getByRole } = subject(defaultProps);

    const input = getByRole('textbox');
    const button = getByRole('button');

    expect(input).not.toBeNull();
    expect(button).not.toBeNull();

    fireEvent.focus(input);
    fireEvent.change(input, { target: { value: '2.5' } });

    expect(mockOnPercentageUpdate).not.toHaveBeenCalled();
    expect(mockOnAutoUpdate).not.toHaveBeenCalled();
    expect(input).toHaveValue('2.5');

    act(() => {
      jest.advanceTimersByTime(300);
    });

    expect(mockOnPercentageUpdate).toHaveBeenCalledWith(new Decimal(0.025));
    expect(mockOnAutoUpdate).toHaveBeenCalledWith(false);

    jest.useRealTimers();
  });

  it('type "." in slippage input will set slippage to zero', () => {
    jest.useFakeTimers();

    const { getByRole } = subject(defaultProps);

    const input = getByRole('textbox');
    const button = getByRole('button');

    expect(input).not.toBeNull();
    expect(button).not.toBeNull();

    fireEvent.focus(input);
    fireEvent.change(input, { target: { value: '.' } });

    expect(mockOnPercentageUpdate).not.toHaveBeenCalled();
    expect(mockOnAutoUpdate).not.toHaveBeenCalled();
    expect(input).toHaveValue('.');

    act(() => {
      jest.advanceTimersByTime(300);
    });

    expect(mockOnPercentageUpdate).toHaveBeenCalledWith(new Decimal(0));
    expect(mockOnAutoUpdate).toHaveBeenCalledWith(false);

    jest.useRealTimers();
  });

  it('click auto button in slippage input will call onAutoUpdate(true)', () => {
    const { getByRole } = subject(defaultProps);

    const input = getByRole('textbox');
    const button = getByRole('button');

    expect(input).not.toBeNull();
    expect(button).not.toBeNull();

    fireEvent.click(button);

    expect(mockOnPercentageUpdate).not.toHaveBeenCalled();
    expect(mockOnAutoUpdate).toHaveBeenCalledWith(true);
  });

  it('type and then set Auto, will call onPercentageUpdate() and onAutoUpdate(true)', () => {
    jest.useFakeTimers();

    const { getByRole } = subject(defaultProps);

    const input = getByRole('textbox');
    const button = getByRole('button');

    expect(input).not.toBeNull();
    expect(button).not.toBeNull();

    fireEvent.change(input, { target: { value: '2.5' } });
    fireEvent.click(button);

    act(() => {
      jest.advanceTimersByTime(300);
    });

    expect(mockOnPercentageUpdate).toHaveBeenCalledWith(new Decimal(0.025));
    expect(mockOnAutoUpdate).toHaveBeenCalledWith(true);

    jest.useRealTimers();
  });

  it('type and then set Auto and then go back to type, will call onPercentageUpdate() and onAutoUpdate(false)', () => {
    jest.useFakeTimers();

    const { getByRole } = subject(defaultProps);

    const input = getByRole('textbox');
    const button = getByRole('button');

    expect(input).not.toBeNull();
    expect(button).not.toBeNull();

    fireEvent.focus(input);
    fireEvent.change(input, { target: { value: '2.5' } });
    fireEvent.blur(input);
    fireEvent.click(button);
    fireEvent.focus(input);
    fireEvent.change(input, { target: { value: '1.5' } });
    fireEvent.blur(input);

    act(() => {
      jest.advanceTimersByTime(300);
    });

    expect(mockOnPercentageUpdate).toHaveBeenCalledWith(new Decimal(0.015));
    expect(mockOnAutoUpdate).toHaveBeenCalledWith(false);

    jest.useRealTimers();
  });
});
