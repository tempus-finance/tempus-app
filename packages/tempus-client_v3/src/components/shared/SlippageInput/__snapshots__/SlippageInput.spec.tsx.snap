// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SlippageInput renders a disabled slippage input 1`] = `
<div
  class="tc__slippage-input"
>
  <div
    class="tc__text-input tc__text-input__disabled"
  >
    <div
      class="tc__text-input__container"
    >
      <div
        style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(122, 122, 122); font-weight: 400; font-family: 'Azeret Mono', monospace; opacity: 1;"
      >
        <input
          class="tc__base-input"
          disabled=""
          id="text-input-3"
          pattern="[0-9,]*[.]?[0-9]{0,2}"
          placeholder="2.00%"
          type="text"
          value=""
        />
      </div>
      <button
        class="tc__btn "
        data-selected="false"
        disabled=""
        type="button"
      >
        <div
          style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(122, 122, 122); font-weight: 400; font-family: 'DM Sans', sans-serif; opacity: 1;"
        >
          Auto
        </div>
      </button>
    </div>
  </div>
</div>
`;

exports[`SlippageInput renders a slippage input 1`] = `
<div
  class="tc__slippage-input"
>
  <div
    class="tc__text-input "
  >
    <div
      class="tc__text-input__container"
    >
      <div
        style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 400; font-family: 'Azeret Mono', monospace; opacity: 1;"
      >
        <input
          class="tc__base-input"
          id="text-input-1"
          pattern="[0-9,]*[.]?[0-9]{0,2}"
          placeholder="2.00%"
          type="text"
          value=""
        />
      </div>
      <button
        class="tc__btn "
        data-selected="false"
        type="button"
      >
        <div
          style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 400; font-family: 'DM Sans', sans-serif; opacity: 1;"
        >
          Auto
        </div>
      </button>
    </div>
  </div>
</div>
`;

exports[`SlippageInput renders a slippage input set to auto 1`] = `
<div
  class="tc__slippage-input"
>
  <div
    class="tc__text-input "
  >
    <div
      class="tc__text-input__container"
    >
      <div
        style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 400; font-family: 'Azeret Mono', monospace; opacity: 1;"
      >
        <input
          class="tc__base-input"
          id="text-input-2"
          pattern="[0-9,]*[.]?[0-9]{0,2}"
          placeholder="2.00%"
          type="text"
          value=""
        />
      </div>
      <button
        class="tc__btn "
        data-selected="true"
        type="button"
      >
        <div
          style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(255, 255, 255); font-weight: 400; font-family: 'DM Sans', sans-serif; opacity: 1;"
        >
          Auto
        </div>
      </button>
    </div>
  </div>
</div>
`;

exports[`SlippageInput renders a slippage input with error 1`] = `
<div
  class="tc__slippage-input"
>
  <div
    class="tc__text-input tc__text-input__error"
  >
    <div
      class="tc__text-input__container"
    >
      <div
        style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 400; font-family: 'Azeret Mono', monospace; opacity: 1;"
      >
        <input
          class="tc__base-input"
          id="text-input-4"
          pattern="[0-9,]*[.]?[0-9]{0,2}"
          placeholder="2.00%"
          type="text"
          value=""
        />
      </div>
      <button
        class="tc__btn "
        data-selected="false"
        type="button"
      >
        <div
          style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 400; font-family: 'DM Sans', sans-serif; opacity: 1;"
        >
          Auto
        </div>
      </button>
    </div>
    <div
      style="font-style: normal; font-size: 12px; line-height: 16px; color: rgb(255, 15, 15); font-weight: 500; font-family: 'DM Sans', sans-serif; opacity: 1;"
    >
      cannot exceed 100%
    </div>
  </div>
</div>
`;
